File: 1231-divide-chocolate/1231-divide-chocolate.java
class Solution {
    public int maximizeSweetness(int[] sweetness, int k) {

        int sum = 0;
        for (int x: sweetness)
            sum+=x;
        
        int left = 0;
        int right = sum;
        
        while (left<right)
        {
            int mid = left+(right-left+1)/2;
            if (ok(sweetness, mid, k))
                left = mid;
            else
                right = mid-1;
        }
        
        return left;
    }
    
    Boolean ok(int[] sweetness, int S, int k) {
        int sum = 0;
        int count = 0;
        for (int i=0; i<sweetness.length; i++)
        {
            sum+=sweetness[i];
            if (sum>=S)
            {
                sum = 0;
                count++;
            }
        }
        return count>=k+1;
    }
}

File: 1231-divide-chocolate/1231-divide-chocolate.py
class Solution:
    def maximizeSweetness(self, sweetness: List[int], k: int) -> int:
        def ok(S: int):
            cnt, amount = 0, 0
            for x in sweetness:
                amount += x
                if amount >= S:
                    cnt += 1
                    amount = 0
            return cnt >= k + 1
        
        l, r = 0, sum(sweetness)
        while l < r:
            m = (r - l + 1) // 2 + l
            if ok(m): l = m
            else: r = m - 1
        return l